packages/flutter_tools/lib/src/commands/precache.dart

```diff
@@ -74,11 +74,14 @@ class PrecacheCommand extends FlutterCommand {
       if (artifact.feature != null && !featureFlags.isEnabled(artifact.feature)) {
         continue;
       }
-      if (boolArg(artifact.name)) {
-        requiredArtifacts.add(artifact);
-      }
       // The `android` flag expands to android_gen_snapshot, android_maven, android_internal_build.
-      if (artifact.name.startsWith('android_') && boolArg('android')) {
+      if (artifact.name.startsWith('android_')) {
+        if (boolArg('android')) {
+          requiredArtifacts.add(artifact);
+        } else {
+          continue;
```

Comment:
This `continue` doesn't look strictly necessary, but it's probably good to keep around as a marker that not including the artifact is intentional. We should probably indicate that with a comment, or maybe pull this logic out into a local helper function of some sort, taking the 'android' string as a parameter---I can imagine we'll want to do something similar with the Fuchsia artifacts at some point since they're already split across the 'fuchsia' and 'flutter_runner' parameters.

Reply:
I started to do this, but the string parsing was getting magical, and would require new artifacts to understand this the "parent_" naming convention:
```dart
  /// Some flags are umbrellas representing multiple artifacts.
  /// This is indicated by prefixing the artifact name with the
  ///
  /// Example: The --android flag expands to android_gen_snapshot, android_maven, android_internal_build.
  /// _parentArtifact("android_internal_build") will return "android".
  String _umbrellaArtifactName(String childArtifact) {
    final List<String> artifactSplit = childArtifact.split('_');
    if (artifactSplit.length >= 2) {
      final String parentArtifact = artifactSplit.first;
      if (argResults.options.contains(parentArtifact)) {
        return parentArtifact;
      }
    }
    return null;
  }
```

I'm going to make this mapping more explicit.